Sources/OAuth2Swift.swift

```diff
@@ -118,32 +124,24 @@ open class OAuth2Swift: OAuthSwift {
             }
         }
 
-        var queryString = "client_id=\(self.consumerKey)"
-        if encodeCallbackURL {
-            queryString += "&redirect_uri=\(callbackURL.absoluteString.urlEncodedString)"
-        } else {
-            queryString += "&redirect_uri=\(callbackURL.absoluteString)"
-        }
-        queryString += "&response_type=\(self.responseType)"
-        if !scope.isEmpty {
-            queryString += "&scope=\(scope)"
-        }
-        if !state.isEmpty {
-            queryString += "&state=\(state)"
-        }
-        for param in parameters {
-            queryString += "&\(param.0)=\(param.1)"
-        }
+        var queryString: String? = ""
 
-        var urlString = self.authorizeUrl
-        urlString += (self.authorizeUrl.contains("?") ? "&" : "?")
+        queryString = queryString?.urlQueryByAppending(parameter: "client_id", value: self.consumerKey)
+        queryString = queryString?.urlQueryByAppending(parameter: "redirect_uri", value: self.encodeCallbackURL ? callbackURL.absoluteString.urlEncoded : callbackURL.absoluteString, encode: self.encodeCallbackURLQuery)
+        queryString = queryString?.urlQueryByAppending(parameter: "response_type", value: self.responseType)
+        queryString = queryString?.urlQueryByAppending(parameter: "scope", value: scope)
+        queryString = queryString?.urlQueryByAppending(parameter: "state", value: state)
 
-        if let encodedQuery = queryString.urlQueryEncoded, let queryURL = URL(string: urlString + encodedQuery) {
-            self.authorizeURLHandler.handle(queryURL)
+        for (name, value) in parameters {
+            queryString = queryString?.urlQueryByAppending(parameter: name, value: "\(value)")
+        }
+
+        if let queryString: String = queryString, let url: URL = URL(string: self.authorizeUrl.urlByAppending(query: queryString)) {
+            self.authorizeURLHandler.handle(url)
             return self
         } else {
             self.cancel() // ie. remove the observer.
-            failure?(OAuthSwiftError.encodingError(urlString: urlString))
+            failure?(OAuthSwiftError.encodingError(urlString: self.authorizeUrl))
```

Comment:
I thought that it might be nice to use [`URLComponents`](https://developer.apple.com/documentation/foundation/urlcomponents) approach with custom encoding, but thought that it might be a radical addition for this PR. Using string extensions and doing low-level url assembly in authorisation method feels wrong. I'm happy to make another PR if you are happy with the approach.